File: app/services/agents.py
"""AgentService: regra de negócio de agentes sobre o AgentRepository (ponto 12).

Criar um agente aqui é o que "abre" o endpoint dele: as rotas resolvem por
(tenant, slug) a cada requisição, então um agente novo responde no mesmo
instante, sem redeploy. A PK é prefixada por tenant (`{tenant}__{slug}`) para
não colidir slug entre tenants (ponto 19).
"""

import logging

from app import rag
from app.domain import AgentConfig
from app.errors import ConflictError, NotFoundError, ValidationError
from app.repositories import AgentRepository
from app.textutil import slugify

log = logging.getLogger("blip-agent.services.agents")


class AgentService:
    def __init__(self, repo: AgentRepository | None = None):
        self._repo = repo or AgentRepository()

    # --- leitura ----------------------------------------------------------- #
    def list_for_tenant(self, tenant_id: str) -> list[AgentConfig]:
        return self._repo.list_for_tenant(tenant_id)

    def list_all(self) -> list[AgentConfig]:
        return self._repo.list_all()

    def get(self, tenant_id: str, slug: str) -> AgentConfig:
        agent = self._repo.get(tenant_id, slug)
        if agent is None:
            raise NotFoundError(f"Agente '{slug}' não encontrado neste tenant.")
        return agent

    # --- escrita ----------------------------------------------------------- #
    def create(self, tenant_id: str, data: dict) -> AgentConfig:
        slug = slugify(data.get("slug") or data["name"], fallback="agente")
        if self._repo.exists(tenant_id, slug):
            raise ConflictError(f"Já existe um agente '{slug}' neste tenant.")
        self._check_external(data.get("product_mode", "none"),
                             data.get("external_products", True))
        agent_id = f"{tenant_id}__{slug}"
        self._repo.insert(agent_id, tenant_id, slug, data)
        log.info("Agente criado: %s (tenant=%s)", agent_id, tenant_id)
        return self._repo.get_by_id(agent_id)

    def update(self, agent: AgentConfig, changes: dict) -> AgentConfig:
        mode = changes.get("product_mode", agent.product_mode)
        ext = changes.get("external_products", agent.external_products)
        self._check_external(mode, ext)
        self._repo.update(agent.id, changes)
        return self._repo.get_by_id(agent.id)

    def delete(self, agent: AgentConfig) -> int:
        """Exclui o agente e a base de conhecimento dele. Produtos saem por
        cascade (FK). Retorna quantos chunks RAG foram removidos."""
        self._repo.delete(agent.id)
        return rag.delete_agent_data(agent.id)

    # --- validação --------------------------------------------------------- #
    @staticmethod
    def _check_external(product_mode: str, external_enabled: bool) -> None:
        if product_mode == "external" and not external_enabled:
            raise ValidationError(
                "Catálogo externo desabilitado para este agente "
                "(feature flag external_products=false)."
            )

File: src/server/services/agents.ts
/**
 * AgentService (porta `app/services/agents.py`): regra de negócio de agentes.
 * Criar um agente "abre" o endpoint dele (resolução por (tenant, slug) a cada
 * requisição). A PK é prefixada por tenant ({tenant}__{slug}).
 */

import type { AgentConfig, ProductMode } from "../domain";
import { ConflictError, NotFoundError, ValidationError } from "../errors";
import { getLogger } from "../logging";
import * as rag from "../rag";
import { AgentRepository } from "../repositories/agents";
import type { AgentCreateData, AgentUpdateData } from "../schemas";
import { allSkillNames } from "../skills";
import { slugify } from "../textutil";

const log = getLogger("blip-agent.services.agents");

export class AgentService {
  constructor(private readonly repo: AgentRepository = new AgentRepository()) {}

  // --- leitura ------------------------------------------------------------- //
  listForTenant(tenantId: string): Promise<AgentConfig[]> {
    return this.repo.listForTenant(tenantId);
  }

  listAll(): Promise<AgentConfig[]> {
    return this.repo.listAll();
  }

  async get(tenantId: string, slug: string): Promise<AgentConfig> {
    const agent = await this.repo.get(tenantId, slug);
    if (agent === null) throw new NotFoundError(`Agente '${slug}' não encontrado neste tenant.`);
    return agent;
  }

  // --- escrita ------------------------------------------------------------- //
  async create(tenantId: string, data: AgentCreateData): Promise<AgentConfig> {
    const slug = slugify(data.slug || data.name, "agente");
    if (await this.repo.exists(tenantId, slug)) {
      throw new ConflictError(`Já existe um agente '${slug}' neste tenant.`);
    }
    AgentService.checkExternal(data.productMode, data.externalProducts);
    AgentService.checkSkills(data.skills);
    const agentId = `${tenantId}__${slug}`;
    await this.repo.insert(agentId, tenantId, slug, data);
    log.info(`Agente criado: ${agentId} (tenant=${tenantId})`);
    return (await this.repo.getById(agentId))!;
  }

  async update(agent: AgentConfig, changes: AgentUpdateData): Promise<AgentConfig> {
    const mode: ProductMode = changes.productMode ?? agent.productMode;
    const ext = changes.externalProducts ?? agent.externalProducts;
    AgentService.checkExternal(mode, ext);
    if (changes.skills !== undefined) AgentService.checkSkills(changes.skills);
    await this.repo.update(agent.id, changes);
    return (await this.repo.getById(agent.id))!;
  }

  /** Exclui o agente e a base de conhecimento dele (produtos saem por cascade). */
  async delete(agent: AgentConfig): Promise<number> {
    await this.repo.delete(agent.id);
    return rag.deleteAgentData(agent.id);
  }

  // --- validação ----------------------------------------------------------- //
  private static checkExternal(productMode: ProductMode, externalEnabled: boolean): void {
    if (productMode === "external" && !externalEnabled) {
      throw new ValidationError(
        "Catálogo externo desabilitado para este agente (feature flag external_products=false).",
      );
    }
  }

  private static checkSkills(skills: string[] | undefined): void {
    if (!skills || skills.length === 0) return;
    const valid = new Set(allSkillNames());
    const unknown = skills.filter((s) => !valid.has(s)).sort();
    if (unknown.length > 0) {
      throw new ValidationError(
        `Skills desconhecidas: ${JSON.stringify(unknown)}. Disponíveis: ${JSON.stringify([...valid].sort())}.`,
      );
    }
  }
}
